db/CSV_Writers/Review_Utils/reviewFunctions.js

```diff
@@ -0,0 +1,46 @@
+const { Promise, path, fs } = require('../packages.js');
+const { twoValProb, threeValProb, generateInBetweenSync } = require('../utils/randomMathGeneration');
+const { primaryRecordNumber, reviewsBigBatchLimiter, _usersRecordsNumber } = require('../config/config.js');
+
+const reviewTableKeys = ["game", "userID", "reviewText", "creationDate", "recommended", "helpfulCount", "notHelpfulCount", "helpfulScore", "funnyCount", "earlyAccess", "awards", "comments"].join(",");
+
+mE = module.exports = {
+  paragraphGetter: (numOfPs, paragraphs) => {
+    return [...new Array(numOfPs)].map(num => paragraphs[generateInBetweenSync(1000)]).join('');
+  },
+
+  reviewBatchGenerator: async () => {
+    const paragraphs = require('../Read_Write_Files/reviews.json');
+    const bigBatchStr = [...new Array(reviewsBigBatchLimiter)].reduce((str) => {
+      const helpfulCount = generateInBetweenSync(100);
+      const notHelpfulCount = generateInBetweenSync(10);
+      const creationDate = (1616535988000 - (generateInBetweenSync(31536000000)));
+      const helpfulScore = helpfulCount / (helpfulCount + notHelpfulCount) * 100;
+      return str += `a,${generateInBetweenSync(primaryRecordNumber)},${generateInBetweenSync(_usersRecordsNumber)},${mE.paragraphGetter(generateInBetweenSync(1), paragraphs)},${creationDate},${twoValProb(.15)},${helpfulCount},${notHelpfulCount},${helpfulScore},${threeValProb(.7, .8)},${threeValProb(.7, .9)},${twoValProb(.9)},${generateInBetweenSync(20)}\r\n`;
+    }, '');
+    fs.writeFileSync("Read_Write_Files/oneMillionReviews.csv", `${reviewTableKeys}\r\n${bigBatchStr}`);
```

Comment:
i know we've been chatting about 'drain' in slack. looks like youre not using here drain here. are you able to avoid draining because youre inserting 1M records ate a time?

Reply:
It seems that 'drain' most specifically applies to a write stream, I was able to remove the inner for loop and generate 1m records (or more) at a time- what you said. Probably if I wanted to remove the outer loop and shoot for the stars in one go then I would have to revisit the drain logic, but for now the process is write 1million (or more) records to a csv file insert, and start over until the query is finished.

Actually I have a sense that awaiting the 'total' or 'str' at the top of this reduce function might actually serve to keep things asynchronous, it certainly doesn't work without that bit. What I understand is that an async function without await returns a promise, so when I return str += 'entry' its returning a promise, that promise resolves before execution continues at the top of the reduce function.